docs/infrastructure/vs-deps-branches.md

```diff
@@ -0,0 +1,132 @@
+2. Your change can't practically be done with some sort of dynamic lightup.
+
+For example, let's say in 16.5 the editor has added a new API to Visual Studio, which requires us to implement a new interface and MEF export it. Because you must implement the interface and MEF export it,
+we need to move to a newer NuGet package that contains the new APIs which will have a higher assembly version, and you must implement the interface which practically can't be done via something like reflection.
+Thus, this change should go into the appropriate -vs-deps branch.
+
+Consider a different case where the editor in 16.5 has added a new API, but we must MEF _import_ it and call a method on it. It's possible to do this via reflection -- you MEF import a raw object,
+and do reflection to call the API accordingly. This is now a trade off we must make. If this is a huge feature that churns thousands of lines of code, but only tiny bit needs to use this new API
+we would generally ask that it goes into master and we use reflection, as the more master and master-vs-deps diverge, the harder the automatic merge becomes. If this is a feature that must use the new API
+everywhere and it'd be vastly more work to do it dynamically than any work we might spend merging stuff, then we'll just have it go into master-vs-deps.
+
+# When do we ever merge from a -vs-deps branch to a regular branch?
+
+Once a version of Visual Studio ships and we are comfortable making that be the new contributor baseline, we do a manual merge from the appropriate -vs-deps branch back. This sometimes means we merge master-vs-deps
+back into master, or sometimes not. If we just shipped 16.4, it's possible that master-vs-deps only depends on stuff in 16.4, so we would merge master-vs-deps back to master. It's also possible that feature work
+has already started on 16.5, and master-vs-deps now contains 16.5 dependent work. The "safe" choice is to merge back just the -vs-deps branch for 16.4, since we know that runs on 16.4 -- it's shipping in it!
+
+# What prevents us from checking stuff what should have gone into a -vs-deps branch into a regular branch?
+
+Integration tests; our integration tests in master are running against released bits, and not random nightly builds. The common case of us moving to a new API in a mainline scenario would break integration tests. Obviously
+if the break is in some untested scenario that won't catch it, but it's also possible it's obscure enough that nobody will ever notice. (If a scenario is broken, but nobody ever tries it, is it really broken?)
+
+# Isn't this all a pain? Can we get rid of the -vs-deps branches?
+
+Maybe, but it potentially replaces one set of problems with another set of problems. So far, the pain of this approach has been the lesser pain, so we've stuck with it. Here's some of the
```

Comment:
Maybe a better question is -- can we create -vs-deps branches on-demand? Does every release need one? Why does every release need one (i.e., is this essential churn, or is there some debt which is causing us to adapt to new APIs constantly).